### PR TITLE
feat(buildFetchUrl): enable optional ids via question mark

### PR DESCRIPTION
**Pull request by PatNeedham**

### CODE REVIEW

**Comment by JAdshead** [comment]
Hi @PatNeedham can you also include an update the readme ?

File: __tests__/helpers/url.spec.js [hunk 1]
```diff
@@ -47,6 +47,18 @@ describe('buildFetchUrl', () => {
     expect(url).toBe('http://api.domain.com/users/123/posts');
   });
 
+  it('should handle empty optional ids by removing extra forward slashes or question marks', () => {
+    const baseUrl = 'http://api.domain.com/users/:userId?/posts/:postId';
+    const url = buildFetchUrl({ url: baseUrl, id: { someId: 'huh' } });
+    expect(url).toBe('http://api.domain.com/users/posts');
+  });
+
+  it('should handle non-empty optional ids', () => {
```
Comment (smackfu) [inline __tests__/helpers/url.spec.js:56]: Can you add a test that has a URL with query params that use a question mark?
e.g. `http://api.domain.com/users?userid=:userId`

File: __tests__/helpers/url.spec.js [hunk 2]
```diff
@@ -47,6 +47,18 @@ describe('buildFetchUrl', () => {
     expect(url).toBe('http://api.domain.com/users/123/posts');
   });
 
+  it('should handle empty optional ids by removing extra forward slashes or question marks', () => {
+    const baseUrl = 'http://api.domain.com/users/:userId?/posts/:postId';
```
Comment (smackfu) [inline __tests__/helpers/url.spec.js:51]: Can you add a test with multiple optional fields?

File: README.md [hunk 3]
```diff
@@ -148,6 +148,7 @@ In an ideal world, your REST API follows all the [right patterns and best practi
 To use the functions that operate on collections, your API and config needs to meet some basic requirements:
 
 * The URL specified on the `fetch` config should return the full collection when the `/:id` is omitted. 
+* URLs can have optional path parameters by including `?`, such as `${process.env.HOST_URL}/users/:userId?/books`.
```
Comment (smackfu) [inline README.md:151]: Maybe:
> Optional path parameters can be indicated by adding a `?` suffix.

File: __tests__/helpers/url.spec.js (hunk 1 above)
Comment (PatNeedham) [reply __tests__/helpers/url.spec.js:56]: This test passes:
```
    const baseUrl = 'http://api.domain.com/users/:userId';
    const url = buildFetchUrl({
      url: baseUrl,
      opts: {
        query: { userId: 'questionable?' },
      },
    });
    expect(url).toBe('http://api.domain.com/users?userId=questionable?');
```
However if I change the baseUrl to end with `/:userId?`, an extra ampersand appears:
image

Would it make sense to include in the readme that marking a path parameter as optional by adding the question mark only works for parameters not at the end of the url? Because for urls with the path param is at the end, it is inherently optional.

File: __tests__/helpers/url.spec.js (hunk 2 above)
Comment (PatNeedham) [reply __tests__/helpers/url.spec.js:51]: Thanks for pointing this out. I had to add the global search flag to the regex in `buildFetchUrl` so it looks like `/\?\//g`. In order to get it to work after adding a test with multiple empty optional ids.

File: README.md (hunk 3 above)
Comment (PatNeedham) [reply README.md]: I updated it to include this suggestion.

File: __tests__/helpers/url.spec.js (hunk 1 above)
Comment (smackfu) [reply __tests__/helpers/url.spec.js:56]: To me it seems like there is still an underlying bug if it's giving surprising results, and I don't think adding it to the README is an effective solution.

Comment (PatNeedham) [reply __tests__/helpers/url.spec.js:56]: Agreed. I guess the `.replace(/\?\//g, '/')` addition converting `?/` matches to `/` also needs to eliminate the trailing `?` if there is one like in this case. But since there's no need to apply that regex to the query params, I'll add it to the end of `replaceUrlParams({ url: remainderUrl, id })` instead

File: __tests__/helpers/url.spec.js [hunk 4]
```diff
@@ -47,6 +47,35 @@ describe('buildFetchUrl', () => {
...
+
+  it('should add query params even when they have question marks', () => {
+    const baseUrl = 'http://api.domain.com/users/:userId';
+    const url = buildFetchUrl({
+      url: baseUrl,
+      opts: {
+        query: { userId: 'questionable?' },
+      },
+    });
+    expect(url).toBe('http://api.domain.com/users?userId=questionable?');
```
[the remark this answers is not in the document]
Comment (PatNeedham) [reply __tests__/helpers/url.spec.js:76]: Honestly not sure, I was mainly trying to think of all the possible cases that could cause it to fail, regardless of how realistic they are.

**Comment by JAdshead** [comment]
@PatNeedham you will need to amend your last commit as its causing the build to fail.

File: __tests__/helpers/url.spec.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (mtomcal) [reply __tests__/helpers/url.spec.js:76]: Gotcha. Okay, Im thinking one more unit test for this case would be good: `http://api.domain.com/users/:user?/comments` and pass:
```
opts: {
  query: { someModifier: 'questionable?' },
},
```
without a `:user` defined and see if:
`expect(url).toBe('http://api.domain.com/users/comments?someModifier=questionable?')`

[the remark this answers is not in the document]
Comment (PatNeedham) [reply __tests__/helpers/url.spec.js:76]: Test fails! 
image
Ampersand is being included instead of the question mark. Well at least initially that was the case. But moving the `.replace(/\?\//g, '/');` to the end of the initial `builtUrl` value (before the `addQueryParams` call) got it to pass, and all remaining tests pass as well.

**Comment by PatNeedham** [comment]
> Add one more test case and rebase and reword commits that dont follow Conventional Commits https://www.conventionalcommits.org/en/v1.0.0/#summary Thanks again!


Since [this change](https://github.com/americanexpress/iguazu-rest/pull/13/commits/b22a7aec7f6f23b187de5a046b8808ac8d6b8dc4#diff-d7bba1b56ab128adc74f01f9d0ad4790) from the first commit is no longer taking place (as the replace is now being called on the `replaceUrlParams` return value), would it be better to squash the two commits into the original instead of rewording the second one?
image

[This one](https://github.com/PatNeedham/iguazu-rest/commit/7e843d5e1d136d31d61f81c4271c8903c00bf4ac) becomes no longer necessary.

**Comment by mtomcal** [comment]
@PatNeedham Yeah definitely go ahead and squash commits in this PR and reword the commit to follow convention. That will fix things. Thanks!
